File: app/src/main/java/com/unmatchedcounter/UnmatchedCharacters.kt
package com.unmatchedcounter

import android.graphics.Color
import androidx.annotation.ColorInt
import androidx.annotation.DrawableRes

/**
 * Playable Unmatched character names for the four life-counter seats.
 * Setup spinners show [NAMES] sorted alphabetically (case-insensitive).
 */
object UnmatchedCharacters {

    /**
     * Single source of truth for display name, life pools, labels, panel art,
     * and setup UI accent (card / spinner / avatar borders).
     */
    private enum class Fighter(
        val displayName: String,
        val segmentLabels: List<String>,
        @DrawableRes val backgroundResId: Int,
        @DrawableRes val avatarResId: Int,
        val startingLifeSegments: List<Int>,
        @ColorInt val setupAccentColorArgb: Int,
    ) {
        GERALT(
            displayName = "Geralt & Dendelion",
            segmentLabels = listOf("GERALT", "DENDELION"),
            backgroundResId = R.drawable.bg_geralt,
            avatarResId = R.drawable.avatar_geralt,
            startingLifeSegments = listOf(16, 5),
            setupAccentColorArgb = Color.parseColor("#309894"),
        ),
        BIGFOOT_JACKALOPE(
            displayName = "Bigfoot & Jackalope",
            segmentLabels = listOf("BIGFOOT", "JACKALOPE"),
            backgroundResId = R.drawable.bg_bigfoot,
            avatarResId = R.drawable.avatar_bigfoot,
            startingLifeSegments = listOf(16, 6),
            setupAccentColorArgb = Color.parseColor("#F0B586"),
        ),
        BRUCE_LEE(
            displayName = "Bruce Lee",
            segmentLabels = listOf("BRUCE LEE"),
            backgroundResId = R.drawable.bg_brucelee,
            avatarResId = R.drawable.avatar_brucelee,
            startingLifeSegments = listOf(14),
            setupAccentColorArgb = Color.parseColor("#ECBD49"),
        ),
        SYNDRA(
            displayName = "Syndra",
            segmentLabels = listOf("SYNDRA"),
            backgroundResId = R.drawable.bg_syndra,
            avatarResId = R.drawable.avatar_syndra,
            startingLifeSegments = listOf(14),
            setupAccentColorArgb = Color.parseColor("#3D25A8"),
        ),
        TASKMASTER(
            displayName = "Taskmaster",
            segmentLabels = listOf("TASKMASTER"),
            backgroundResId = R.drawable.bg_taskmaster,
            avatarResId = R.drawable.avatar_taskmaster,
            startingLifeSegments = listOf(16),
            setupAccentColorArgb = Color.parseColor("#FF8B33"),
        ),
        MULDOON_WORKERS(
            displayName = "Muldoon & Workers",
            segmentLabels = listOf("MULDOON"),
            backgroundResId = R.drawable.bg_muldoon,
            avatarResId = R.drawable.avatar_muldoon,
            startingLifeSegments = listOf(14),
            setupAccentColorArgb = Color.parseColor("#E39139"),
        ),
        CHUPACABRAS(
            displayName = "Chupacabras",
            segmentLabels = listOf("CHUPACABRAS"),
            backgroundResId = R.drawable.bg_chupacabras,
            avatarResId = R.drawable.avatar_chupacabras,
            startingLifeSegments = listOf(14),
            setupAccentColorArgb = Color.parseColor("#C0C565"),
        ),
        RAPTORS(
            displayName = "Raptors",
            segmentLabels = listOf("BLUE", "ECHO", "CHARLIE"),
            backgroundResId = R.drawable.bg_raptors,
            avatarResId = R.drawable.avatar_raptors,
            startingLifeSegments = listOf(7, 7, 7),
            setupAccentColorArgb = Color.parseColor("#AF9E49"),
        ),
        EREDIN(
            displayName = "Eredin",
            segmentLabels = listOf("EREDIN"),
            backgroundResId = R.drawable.bg_eredin,
            avatarResId = R.drawable.avatar_eredin,
            startingLifeSegments = listOf(14),
            setupAccentColorArgb = Color.parseColor("#2E302D"),
        ),
        BULLSEYE(
            displayName = "Bullseye",
            segmentLabels = listOf("BULLSEYE"),
            backgroundResId = R.drawable.bg_bullseye,
            avatarResId = R.drawable.avatar_bullseye,
            startingLifeSegments = listOf(14),
            setupAccentColorArgb = Color.parseColor("#5876A1"),
        ),
        HOUDINI_BESS(
            displayName = "Houdini & Bess",
            segmentLabels = listOf("HOUDINI", "BESS"),
            backgroundResId = R.drawable.bg_houdini,
            avatarResId = R.drawable.avatar_houdini,
            startingLifeSegments = listOf(14, 5),
            setupAccentColorArgb = Color.parseColor("#C6AD5A"),
        ),
        DAREDEVIL(
            displayName = "Daredevil",
            segmentLabels = listOf("DAREDEVIL"),
            backgroundResId = R.drawable.bg_daredevil,
            avatarResId = R.drawable.avatar_daredevil,
            startingLifeSegments = listOf(17),
            setupAccentColorArgb = Color.parseColor("#770A08"),
        ),
        LOKI(
            displayName = "Loki",
            segmentLabels = listOf("LOKI"),
            backgroundResId = R.drawable.bg_loki,
            avatarResId = R.drawable.avatar_loki,
            startingLifeSegments = listOf(16),
            setupAccentColorArgb = Color.parseColor("#354932"),
        ),
        SHERLOCK_WATSON(
            displayName = "Sherlock & Dr.Watson",
            segmentLabels = listOf("SHERLOCK", "DR. WATSON"),
            backgroundResId = R.drawable.bg_sherlock,
            avatarResId = R.drawable.avatar_sherlock,
            startingLifeSegments = listOf(16, 8),
            setupAccentColorArgb = Color.parseColor("#FFC03F"),
        ),
        ELEKTRA(
            displayName = "Elektra",
            segmentLabels = listOf("ELEKTRA"),
            backgroundResId = R.drawable.bg_elektra,
            avatarResId = R.drawable.avatar_elektra,
            startingLifeSegments = listOf(8),
            setupAccentColorArgb = Color.parseColor("#770A08"),
        ),
        ZED(
            displayName = "Zed",
            segmentLabels = listOf("ZED"),
            backgroundResId = R.drawable.bg_zed,
            avatarResId = R.drawable.avatar_zed,
            startingLifeSegments = listOf(15),
            setupAccentColorArgb = Color.parseColor("#770A08"),
        ),
        KING_ARTHUR_MERLIN(
            displayName = "Arthur & Merlin",
            segmentLabels = listOf("KING ARTHUR", "MERLIN"),
            backgroundResId = R.drawable.bg_kingarthur,
            avatarResId = R.drawable.avatar_arthur,
            startingLifeSegments = listOf(18, 7),
            setupAccentColorArgb = Color.parseColor("#2E302D"),
        ),
        MEDUSA_HARPIES(
            displayName = "Medusa",
            segmentLabels = listOf("MEDUSA"),
            backgroundResId = R.drawable.bg_medusa,
            avatarResId = R.drawable.avatar_medusa,
            startingLifeSegments = listOf(16),
            setupAccentColorArgb = Color.parseColor("#354932"),
        ),
        ALICE_JABBERWOCK(
            displayName = "Alice & Jabberwock",
            segmentLabels = listOf("ALICE", "JABBERWOCK"),
            backgroundResId = R.drawable.bg_alicia,
            avatarResId = R.drawable.avatar_alice,
            startingLifeSegments = listOf(13, 8),
            setupAccentColorArgb = Color.parseColor("#5876A1"),
        ),
        SINBAD_PORTER(
            displayName = "Sinbad & The Porter",
            segmentLabels = listOf("SINBAD", "PORTER"),
            backgroundResId = R.drawable.bg_simbad,
            avatarResId = R.drawable.avatar_simbad,
            startingLifeSegments = listOf(15, 6),
            setupAccentColorArgb = Color.parseColor("#F0B586"),
        ),
    }

    val NAMES: List<String> = Fighter.entries.map { it.displayName }

    private val SEGMENT_LABELS_BY_NAME: Map<String, List<String>> =
        Fighter.entries.associate { it.displayName to it.segmentLabels }

    private val BACKGROUND_DRAWABLE_RES_BY_NAME: Map<String, Int> =
        Fighter.entries.associate { it.displayName to it.backgroundResId }

    private val AVATAR_DRAWABLE_RES_BY_NAME: Map<String, Int> =
        Fighter.entries.associate { it.displayName to it.avatarResId }

    private val STARTING_LIFE_SEGMENTS_BY_NAME: Map<String, List<Int>> =
        Fighter.entries.associate { it.displayName to it.startingLifeSegments }

    private val SETUP_ACCENT_ARGB_BY_NAME: Map<String, Int> =
        Fighter.entries.associate { it.displayName to it.setupAccentColorArgb }

    init {
        for (f in Fighter.entries) {
            val pools = f.startingLifeSegments.size
            val labels = f.segmentLabels.size
            require(pools == labels) {
                "Fighter ${f.name}: distinto nº de pools y etiquetas ($pools vs $labels)"
            }
            require(f.startingLifeSegments.isNotEmpty()) { "Fighter ${f.name}: sin pools de vida" }
            require(f.startingLifeSegments.size <= 3) { "Fighter ${f.name}: máximo 3 pools" }
        }
    }

    /**
     * Border accent for setup card, spinner, and avatar for this fighter.
     */
    @ColorInt
    fun setupAccentColorFor(characterName: String): Int {
        require(characterName in NAMES) { "Unknown character: $characterName" }
        return SETUP_ACCENT_ARGB_BY_NAME.getValue(characterName)
    }

    /**
     * @return a non-empty list (length 1..3) of starting life values, one per pool.
     */
    fun startingLifeSegmentsFor(characterName: String): List<Int> {
        require(characterName in NAMES) { "Unknown character: $characterName" }
        return STARTING_LIFE_SEGMENTS_BY_NAME.getValue(characterName).map { it }
    }

    /**
     * @return one display string per life pool, same order and length as [startingLifeSegmentsFor].
     */
    fun segmentLabelsFor(characterName: String): List<String> {
        require(characterName in NAMES) { "Unknown character: $characterName" }
        return SEGMENT_LABELS_BY_NAME.getValue(characterName).map { it }
    }

    /** Round avatar shown on the setup card for this fighter. */
    @DrawableRes
    fun avatarDrawableResId(characterName: String): Int {
        require(characterName in NAMES) { "Unknown character: $characterName" }
        return AVATAR_DRAWABLE_RES_BY_NAME.getValue(characterName)
    }

    /** Panel art behind life digits for this fighter (one drawable per character name). */
    @DrawableRes
    fun backgroundDrawableResId(characterName: String, playerIdForFallback: Int): Int {
        val name = characterName.takeIf { it in NAMES }
            ?: NAMES.getOrElse(playerIdForFallback.coerceIn(0, NAMES.lastIndex)) { NAMES.first() }
        return BACKGROUND_DRAWABLE_RES_BY_NAME.getValue(name)
    }
}

File: ios/UnmatchedCounter/Model/UnmatchedCharacters.swift
import UIKit

/// Catalog of playable Unmatched characters for the four life-counter seats.
/// Mirrors `UnmatchedCharacters.kt` from the Android source: each fighter has
/// 1..3 life pools, per-pool labels, panel art, avatar art, and a UI accent.
enum UnmatchedCharacters {

    private struct Fighter {
        let displayName: String
        let segmentLabels: [String]
        let backgroundAsset: String
        let avatarAsset: String
        let startingLifeSegments: [Int]
        let setupAccentHex: UInt32
    }

    private static let fighters: [Fighter] = [
        Fighter(displayName: "Geralt & Dendelion",
                segmentLabels: ["GERALT", "DENDELION"],
                backgroundAsset: "bg_geralt", avatarAsset: "avatar_geralt",
                startingLifeSegments: [16, 5], setupAccentHex: 0x309894),
        Fighter(displayName: "Bigfoot & Jackalope",
                segmentLabels: ["BIGFOOT", "JACKALOPE"],
                backgroundAsset: "bg_bigfoot", avatarAsset: "avatar_bigfoot",
                startingLifeSegments: [16, 6], setupAccentHex: 0xF0B586),
        Fighter(displayName: "Bruce Lee",
                segmentLabels: ["BRUCE LEE"],
                backgroundAsset: "bg_brucelee", avatarAsset: "avatar_brucelee",
                startingLifeSegments: [14], setupAccentHex: 0xECBD49),
        Fighter(displayName: "Syndra",
                segmentLabels: ["SYNDRA"],
                backgroundAsset: "bg_syndra", avatarAsset: "avatar_syndra",
                startingLifeSegments: [14], setupAccentHex: 0x3D25A8),
        Fighter(displayName: "Taskmaster",
                segmentLabels: ["TASKMASTER"],
                backgroundAsset: "bg_taskmaster", avatarAsset: "avatar_taskmaster",
                startingLifeSegments: [16], setupAccentHex: 0xFF8B33),
        Fighter(displayName: "Muldoon & Workers",
                segmentLabels: ["MULDOON"],
                backgroundAsset: "bg_muldoon", avatarAsset: "avatar_muldoon",
                startingLifeSegments: [14], setupAccentHex: 0xE39139),
        Fighter(displayName: "Chupacabras",
                segmentLabels: ["CHUPACABRAS"],
                backgroundAsset: "bg_chupacabras", avatarAsset: "avatar_chupacabras",
                startingLifeSegments: [14], setupAccentHex: 0xC0C565),
        Fighter(displayName: "Raptors",
                segmentLabels: ["BLUE", "ECHO", "CHARLIE"],
                backgroundAsset: "bg_raptors", avatarAsset: "avatar_raptors",
                startingLifeSegments: [7, 7, 7], setupAccentHex: 0xAF9E49),
        Fighter(displayName: "Eredin",
                segmentLabels: ["EREDIN"],
                backgroundAsset: "bg_eredin", avatarAsset: "avatar_eredin",
                startingLifeSegments: [14], setupAccentHex: 0x2E302D),
        Fighter(displayName: "Bullseye",
                segmentLabels: ["BULLSEYE"],
                backgroundAsset: "bg_bullseye", avatarAsset: "avatar_bullseye",
                startingLifeSegments: [14], setupAccentHex: 0x5876A1),
        Fighter(displayName: "Houdini & Bess",
                segmentLabels: ["HOUDINI", "BESS"],
                backgroundAsset: "bg_houdini", avatarAsset: "avatar_houdini",
                startingLifeSegments: [14, 5], setupAccentHex: 0xC6AD5A),
        Fighter(displayName: "Daredevil",
                segmentLabels: ["DAREDEVIL"],
                backgroundAsset: "bg_daredevil", avatarAsset: "avatar_daredevil",
                startingLifeSegments: [17], setupAccentHex: 0x770A08),
        Fighter(displayName: "Loki",
                segmentLabels: ["LOKI"],
                backgroundAsset: "bg_loki", avatarAsset: "avatar_loki",
                startingLifeSegments: [16], setupAccentHex: 0x354932),
        Fighter(displayName: "Sherlock & Dr.Watson",
                segmentLabels: ["SHERLOCK", "DR. WATSON"],
                backgroundAsset: "bg_sherlock", avatarAsset: "avatar_sherlock",
                startingLifeSegments: [16, 8], setupAccentHex: 0xFFC03F),
        Fighter(displayName: "Elektra",
                segmentLabels: ["ELEKTRA"],
                backgroundAsset: "bg_elektra", avatarAsset: "avatar_elektra",
                startingLifeSegments: [8], setupAccentHex: 0x770A08),
        Fighter(displayName: "Zed",
                segmentLabels: ["ZED"],
                backgroundAsset: "bg_zed", avatarAsset: "avatar_zed",
                startingLifeSegments: [15], setupAccentHex: 0x770A08),
        Fighter(displayName: "Arthur & Merlin",
                segmentLabels: ["KING ARTHUR", "MERLIN"],
                backgroundAsset: "bg_kingarthur", avatarAsset: "avatar_arthur",
                startingLifeSegments: [18, 7], setupAccentHex: 0x2E302D),
        Fighter(displayName: "Medusa",
                segmentLabels: ["MEDUSA"],
                backgroundAsset: "bg_medusa", avatarAsset: "avatar_medusa",
                startingLifeSegments: [16], setupAccentHex: 0x354932),
        Fighter(displayName: "Alice & Jabberwock",
                segmentLabels: ["ALICE", "JABBERWOCK"],
                backgroundAsset: "bg_alicia", avatarAsset: "avatar_alice",
                startingLifeSegments: [13, 8], setupAccentHex: 0x5876A1),
        Fighter(displayName: "Sinbad & The Porter",
                segmentLabels: ["SINBAD", "PORTER"],
                backgroundAsset: "bg_simbad", avatarAsset: "avatar_simbad",
                startingLifeSegments: [15, 6], setupAccentHex: 0xF0B586),
    ]

    private static let byName: [String: Fighter] = {
        var d: [String: Fighter] = [:]
        for f in fighters {
            precondition(f.startingLifeSegments.count == f.segmentLabels.count,
                         "Fighter \(f.displayName) has mismatched pools/labels")
            precondition(!f.startingLifeSegments.isEmpty && f.startingLifeSegments.count <= LifeSegmentLimits.maxLifeSegments,
                         "Fighter \(f.displayName) must have 1..\(LifeSegmentLimits.maxLifeSegments) pools")
            d[f.displayName] = f
        }
        return d
    }()

    /// All character names in catalog order.
    static let names: [String] = fighters.map { $0.displayName }

    /// All character names sorted alphabetically (case-insensitive), for pickers.
    static let sortedNames: [String] = names.sorted {
        $0.compare($1, options: .caseInsensitive) == .orderedAscending
    }

    static func setupAccentColor(for characterName: String) -> UIColor {
        guard let f = byName[characterName] else {
            preconditionFailure("Unknown character: \(characterName)")
        }
        return UIColor(rgbHex: f.setupAccentHex)
    }

    static func startingLifeSegments(for characterName: String) -> [Int] {
        guard let f = byName[characterName] else {
            preconditionFailure("Unknown character: \(characterName)")
        }
        return f.startingLifeSegments
    }

    static func segmentLabels(for characterName: String) -> [String] {
        guard let f = byName[characterName] else {
            preconditionFailure("Unknown character: \(characterName)")
        }
        return f.segmentLabels
    }

    static func avatarImage(for characterName: String) -> UIImage? {
        guard let f = byName[characterName] else { return nil }
        return UIImage(named: f.avatarAsset)
    }

    /// Falls back to a deterministic art slot based on the player id when the
    /// character is unknown (mirrors the Android implementation).
    static func backgroundImage(for characterName: String, playerIdForFallback: Int) -> UIImage? {
        let resolvedName: String
        if byName[characterName] != nil {
            resolvedName = characterName
        } else if !names.isEmpty {
            let idx = max(0, min(playerIdForFallback, names.count - 1))
            resolvedName = names[idx]
        } else {
            return nil
        }
        guard let f = byName[resolvedName] else { return nil }
        return UIImage(named: f.backgroundAsset)
    }
}

extension UIColor {
    /// Build a UIColor from a 24-bit RGB packed integer (0xRRGGBB).
    convenience init(rgbHex: UInt32, alpha: CGFloat = 1.0) {
        let r = CGFloat((rgbHex >> 16) & 0xFF) / 255.0
        let g = CGFloat((rgbHex >> 8) & 0xFF) / 255.0
        let b = CGFloat(rgbHex & 0xFF) / 255.0
        self.init(red: r, green: g, blue: b, alpha: alpha)
    }
}
